### PR TITLE
APEXMALHAR-2034 Adding new Avro Module to encapsulate Container File to Avro GenericRecord to POJO transformations

### PR DESCRIPTION
**Pull request by SaumyaMohan**
Adding link to closed PR: https://github.com/apache/apex-malhar/pull/665

New PR is created to ensure forked branch starts with JIRA name

### CODE REVIEW

**Comment by tweise** [comment]
Why is a new Maven project added when goal is to just add the Avro operator/module that combines existing operators? Wouldn't it be more appropriate to add that class to the existing package? 

@sandeshh Adding the new project will require more work, including history preserving renames, proper dependency management, documentation, meaningful test coverage for the CI and other things that justify such invasive change. 

Most importantly, I think that it would warrant an upfront discussion. What we don't need is a lot of code moved around when the net benefit is small.

Also, will this module work with partitioning? Let's say Kafka -> Avro -> POJO all in a parallel partition?

**Comment by SaumyaMohan** [comment]
@tweise Creating a new maven module for Avro is in accordance with the JIRA https://issues.apache.org/jira/browse/APEXMALHAR-1843. Since a new 'apex module' is being created for Avro, we decided to add it strategically in accordance with the above JIRA.

We have ensured that the files that are moved to the new location retain the GIT history.
Unit test for the new 'apex module' has been added
Current users referring to the old location of Avro operators will continue to be able to use them as the operators in old location are not deleted but marked deprecated, and they extend the operators in the new location. This has been tested by the running a sample app using old location operators.

**Comment by sandeshh** [comment]
Following combination, Kafka -> Avro -> POJO, is not supported. Avro operator is an input operator which extends AbstractFileInputOperator, so parallel partitioning behavior is same as AbstractFileInputOperator's behavior.

**Comment by SaumyaMohan** [comment]
I've updated the JIRA https://issues.apache.org/jira/browse/APEXMALHAR-2034 with all the tests and checks done to ensure proper dependencies, expected behavior of Avro module, and backward compatibility.

File: avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    }
+  }
+
+  private void writeAvroFile(File outputFile) throws IOException
+  {
+
+    DatumWriter<GenericRecord> datumWriter = new GenericDatumWriter<GenericRecord>(
+        new Schema.Parser().parse(AVRO_SCHEMA));
+
+    DataFileWriter<GenericRecord> dataFileWriter = new DataFileWriter<GenericRecord>(datumWriter);
```
Comment (sandeshh) [inline avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java]: Use Try-with-resources structure

File: avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  @Rule
+  public TestMeta testMeta = new TestMeta();
+
+  private void createAvroInput(int cnt)
```
Comment (sandeshh) [inline avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java]: Code duplication? please create util functions

File: avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+      rec.put("orderId", cnt * 1L);
+      rec.put("customerId", cnt * 2);
+      rec.put("total", cnt * 1.5);
+      rec.put("customerName", "*" + cnt + "*");
+      cnt--;
+      recordList.add(rec);
+    }
+  }
+
+  private void writeAvroFile(File outputFile) throws IOException
```
Comment (sandeshh) [inline avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java]: Where is the cleanup of the files?

**Comment by tweise** [comment]
@SaumyaMohan addition of the composite Avro operator does not require the creation of a new Maven project and code refactor. You can separate the two if you like, but the project changes need to be discussed on the mailing list first, @sandeshh should be aware of that.

File: avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java (hunk 3 above)
Comment (SaumyaMohan) [reply avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java]: the file cleanup is done by the overloaded finished() method

File: avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java (hunk 1 above)
Comment (SaumyaMohan) [reply avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java]: fixed

File: avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java (hunk 2 above)
Comment (sandeshh) [reply avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java]: Let us not worry about refactoring in this jira. You can create a separate jira.

**Comment by sandeshh** [comment]
Will merge it tomorrow.

**Comment by tweise** [comment]
-1 on the PR as it is. I see no justification to create a new Maven project for this change. You can add the composite operator to the existing package.

**Comment by SaumyaMohan** [comment]
@tweise thanks for your feedback. I just want to understand why a strategic move to a new package as per existing JIRA https://issues.apache.org/jira/browse/APEXMALHAR-1843 is being discouraged. I understand the functional change is small but through this PR we have tried to not only improve the functionality but also take this opportunity to improve the project structure which was already encouraged in the APEXMALHAR-1843 (unless that JIRA should not be referred to by developers).

We have taken proper steps to ensure file history and backward compatibility is maintained.

We have also started an email chain sending our justification for this move. 
Please could you respond with your concerns on that email so all Apache Apex developers are on the same page on what approach to take for development tasks?

**Comment by tweise** [comment]
Yes, we can continue discussing this on the email thread. It is the recommended process for changes of this extend, to avoid contributors wasting efforts.

**Comment by SaumyaMohan** [comment]
@tweise if I've already taken the initiative to make the strategic structural change (as per existing JIRA as mentioned above), please could you elaborate what's the technical reason why this should not be done?

I just want to understand if the move to new package is something not considered strategic or it should not be done right now.

**Comment by tweise** [comment]
@SaumyaMohan you should really use the mailing list to discuss this ahead of taking up the work, that's part of the collaboration process.

It is correct that the project structure needs to become more modular and in principle this is the right move. But it is not just about moving around existing code. Contrib generally implies lack of maturity/stability and when I tried to use the Avro operator in the past, it was frankly not ready for prime time. Promotion to a separate module needs to come with a base line in readiness, otherwise the entire move is just cosmetical. So before taking up this kind of initiative, you should make the case why it makes sense at this time and propose what needs to be done to provide adequate value to a user. I also agree that the JIRA isn't very helpful, it was created a long time ago and could benefit from some more context based on where the project is today.

As already stated, you could break up your PR into addressing the problem that you originally set out to solve (add new operator) as one change and new project creation separately.

File: avro/src/main/java/org/apache/apex/malhar/avro/AvroFileInputOperator.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+import com.datatorrent.api.DefaultOutputPort;
+import com.datatorrent.lib.io.fs.AbstractFileInputOperator;
+
+/**
+ * <p>
+ * Avro File Input Operator
+ * </p>
+ * A specific implementation of the AbstractFileInputOperator to read Avro
+ * container files.<br>
+ * No need to provide schema,its inferred from the file<br>
```
Comment (tweise) [inline avro/src/main/java/org/apache/apex/malhar/avro/AvroFileInputOperator.java]: It's not "inferred", the schema is stored in the file and that is standard Avro.

File: avro/src/main/java/org/apache/apex/malhar/avro/AvroFileInputOperator.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        recordCount++;
+        return record;
+      }
+    } catch (AvroRuntimeException are) {
+      LOG.error("Exception in parsing record for file - " + super.currentFile + " at offset - " + offset, are);
+      if (errorRecordsPort.isConnected()) {
+        errorRecordsPort.emit("FileName:" + super.currentFile + ", Offset:" + offset);
+      }
+      errorCount++;
+      throw new AvroRuntimeException(are);
```
Comment (tweise) [inline avro/src/main/java/org/apache/apex/malhar/avro/AvroFileInputOperator.java]: why is the original exception not thrown

File: avro/pom.xml [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    </build>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.codehaus.janino</groupId>
+            <artifactId>janino</artifactId>
+            <version>2.7.8</version>
+        </dependency>
+        <dependency>
+            <groupId>org.python</groupId>
```
Comment (tweise) [inline avro/pom.xml]: why is this needed?

File: avro/pom.xml [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+                        <goals>
+                            <goal>run</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <!-- generate javdoc -->
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-javadoc-plugin</artifactId>
```
Comment (tweise) [inline avro/pom.xml]: why all of this clutter? what is it needed for?

File: avro/pom.xml [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.apex</groupId>
+            <artifactId>apex-common</artifactId>
+            <version>${apex.core.version}</version>
+            <type>jar</type>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-annotations</artifactId>
```
Comment (tweise) [inline avro/pom.xml]: why are these dependencies declared here?

File: avro/pom.xml [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>2.7.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.avro</groupId>
+            <artifactId>avro</artifactId>
+            <version>1.7.4</version>
+            <exclusions>
```
Comment (tweise) [inline avro/pom.xml]: why the exclusions?

File: avro/src/main/java/org/apache/apex/malhar/avro/AvroFileToPojoModule.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+ * to ensure exactly once semantics with a HDFS backed WAL.
+ *
+ * @displayName AvroFileToPojoModule
+ * @category Input
+ * @tags fs, file,avro, input operator, generic record, pojo
+ *
+ * @since
+ */
+
+public class AvroFileToPojoModule implements Module
```
Comment (tweise) [inline avro/src/main/java/org/apache/apex/malhar/avro/AvroFileToPojoModule.java]: What value does "Module" in the name add?

File: avro/src/main/java/org/apache/apex/malhar/avro/AvroFileToPojoModule.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  public final transient ProxyOutputPort<GenericRecord> errorPort = new ProxyOutputPort<>();
+  //output ports from AvroFileInputOperator
+  public final transient ProxyOutputPort<String> completedAvroFilesPort = new ProxyOutputPort<>();
+  public final transient ProxyOutputPort<String> avroErrorRecordsPort = new ProxyOutputPort<>();
+
+  private AvroFileInputOperator avroFileInputOperator = new AvroFileInputOperator();
+  Class<?> pojoClass = null;
+
+  @Override
+  public void populateDAG(DAG dag, Configuration configuration)
```
Comment (tweise) [inline avro/src/main/java/org/apache/apex/malhar/avro/AvroFileToPojoModule.java]: I asked before how parallel partitioning works with this?

File: avro/src/main/java/org/apache/apex/malhar/avro/AvroRecordHelper.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+      convertedValue = convertAndResolveUnionToPrimitive(schema, key, value);
+    } else {
+      convertedValue = convertValueToAvroPrimitive(type, key, value);
+    }
+
+    return convertedValue;
+
+  }
+
+  private static Object convertValueToAvroPrimitive(Type type, String key, String value) throws ParseException
```
Comment (tweise) [inline avro/src/main/java/org/apache/apex/malhar/avro/AvroRecordHelper.java]: Why is there an assumption that every value is String, when conversion is from/to POJO. This also does not support all types that are supported in Avro? Latest Avro version is 1.8.x

File: avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+ * <p>
+ * In this class the emitTuples method is called twice to process the first
+ * input, since on begin window 0 the operator is setup & stream is initialized.
+ * The platform calls the emitTuples method in the successive windows
+ * </p>
+ */
+public class AvroFileToPojoModuleTest
+{
+
+  private static final String AVRO_SCHEMA = "{\"namespace\":\"abc\"," + ""
```
Comment (tweise) [inline avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java]: why is this when schemas can be stored as JSON file resources

File: avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java (hunk 2 above)
Comment (tweise) [reply avro/src/test/java/org/apache/apex/malhar/avro/AvroFileToPojoModuleTest.java]: we actually should worry about this and other issues when promoting an operator to a new module

**Review by tweise — CHANGES_REQUESTED** [review]
I just looked through parts of this PR and there are quite a few issues with it. How can this be declared as ready for merge? As discussed, it is better to focus on one thing and do that right then dealing with a mountain of inherited code issues while attempting to address unrelated issues at once.

File: contrib/pom.xml [hunk 14]
```diff
@@ -678,5 +678,10 @@
       <version>1.3.0</version>
       <optional>true</optional>
     </dependency>
+    <dependency>
```
Comment (tweise) [inline contrib/pom.xml]: This doesn't belong here. Not everyone using contrib wants to pull in the Avro module.

File: avro/src/main/java/org/apache/apex/malhar/avro/AvroFileInputOperator.java (hunk 5 above)
Comment (sandeshh) [reply avro/src/main/java/org/apache/apex/malhar/avro/AvroFileInputOperator.java]: Few of these comments are on the existing code, which we were not planning to change as a part of this PR. We wanted to improve this operator step by step. Anyways, now it is up to @SaumyaMohan to decide whether she wants to continue with this PR.

**Comment by SaumyaMohan** [comment]
@tweise most of the issues you pointed out are in the existing code which I am not changing. It seems it is easier to make the code change in the old contrib location so I will go ahead and do that and not move it to new location for this PR.

Regarding your question on how parallel partitioning will work for Avro, @sandeshh already replied "Avro operator is an input operator which extends AbstractFileInputOperator, so parallel partitioning behavior is same as AbstractFileInputOperator's behavior."

**Comment by SaumyaMohan** [comment]
Creating a new Pull Request with only the code changes and no file move
